backend/dataall/modules/worksheets/services/worksheet_query_result_service.py

```diff
@@ -0,0 +1,139 @@
+        from sqlalchemy.orm import Session
+        from mypy_boto3_s3.client import S3Client
+    except ImportError:
+        print('skipping type checks as stubs are not installed')
+        S3Client = None
+        Session = None
+
+
+class WorksheetQueryResultService:
+    SupportedFormats = {'csv', 'xlsx'}
+
+    @staticmethod
+    def validate_input(data):
+        if not data:
+            raise exceptions.InvalidInput('data', data, 'input is required')
+        if not data.get('athenaQueryId'):
+            raise exceptions.RequiredParameter('athenaQueryId')
+        if not data.get('fileFormat'):
+            raise exceptions.RequiredParameter('fileFormat')
+        if data.get('fileFormat', '').lower() not in WorksheetQueryResultService.SupportedFormats:
+            raise exceptions.InvalidInput(
+                'fileFormat', data.get('fileFormat'), ', '.join(WorksheetQueryResultService.SupportedFormats)
+            )
+
+    @staticmethod
+    def get_output_bucket(session: 'Session', environment_uri: str) -> str:
+        environment = EnvironmentService.get_environment_by_uri(session, environment_uri)
+        bucket = environment.EnvironmentDefaultBucketName
+        return bucket
+
+    @staticmethod
+    def create_query_result(
+        environment_bucket: str, athena_workgroup: str, worksheet_uri: str, region: str, aws_account_id: str, data: dict
+    ) -> WorksheetQueryResult:
+        sql_query_result = WorksheetQueryResult(
+            worksheetUri=worksheet_uri,
+            AthenaQueryId=data.get('athenaQueryId'),
+            fileFormat=data.get('fileFormat'),
+            OutputLocation=f's3://{environment_bucket}/athenaqueries/{athena_workgroup}/',
+            region=region,
+            AwsAccountId=aws_account_id,
+            queryType='data',
+        )
+        return sql_query_result
+
+    @staticmethod
+    def get_file_key(
+        workgroup: str, query_id: str, file_format: str = 'csv', athena_queries_dir: str = 'athenaqueries'
+    ) -> str:
+        return f'{athena_queries_dir}/{workgroup}/{query_id}.{file_format}'
+
+    @staticmethod
+    def convert_csv_to_xlsx(csv_data) -> io.BytesIO:
+        wb = Workbook()
+        ws = wb.active
+        csv_reader = csv.reader(csv_data.splitlines())
+        for row in csv_reader:
+            ws.append(row)
+
+        excel_buffer = io.BytesIO()
+        wb.save(excel_buffer)
+        excel_buffer.seek(0)
+        return excel_buffer
+
+    @staticmethod
+    def handle_xlsx_format(output_bucket: str, file_key: str) -> bool:
+        aws_region_name = os.getenv('AWS_REGION_NAME', 'eu-west-1')
+        file_name, _ = file_key.split('.')
+        csv_data = S3_client.get_object(region=aws_region_name, bucket=output_bucket, key=f'{file_name}.csv')
+        excel_buffer = WorksheetQueryResultService.convert_csv_to_xlsx(csv_data)
+        S3_client.put_object(region=aws_region_name, bucket=output_bucket, key=file_key, body=excel_buffer)
+        return True
+
```

Comment:
external functions used by resolvers should be decorated with permission checks. In other words, we need to ensure from the server-side that the API call to download_sql_query_results can only be executed by users with permissions.

```
    @staticmethod
    @ResourcePolicyService.has_resource_permission(RUN_ATHENA_QUERY)
    def download_sql_query_result(session, uri, data: dict = None):
```

Reply:
I think the `RUN_ATHENA_QUERY` permission can be reused. If a user has permissions to run the query, then he should have permissions to download the results. It can be implemented later if needed, but do you think there will be a case in which we need to distinguish between users that can run queries and users that can download the results?

Reply:
Right now! I do not see any such use-case for distinguishing the user for running the query and downloading the result.
We can safely proceed ahead with `RUN_ATHENA_QUERY` permission

Reply:
At the end there is a dedicated permission applied on the worksheet, but I think it is good. We can resolve